chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,144 @@
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+const { getColumns, addColumn, removeColumn, getColumn } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOption} ItemTreeColumnOption
+ */
+
+class ItemTreeManager {
+    _observerAdded = false;
+    constructor() {
+    }
+
+    /** 
+     * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+     * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} option
+     * @async
+     * @returns {boolean} Although it's async, resolving does not promise the item trees are updated.
+     * @throws {Error} If the column option is missing required properties
+     * @throws {Error} If the column is already registered
+     */
+    async registerColumn(option) {
+        const success = addColumn(option);
+        if (!success) {
+            return false;
+        }
+        this._ensureObserverAdded();
+        await this._reset();
+        return true;
+    }
+
+    /**
+     * Unregister a custom column
+     * @param {string} dataKey - The dataKey of the column to unregister
+     * @async
+     * @returns {boolean} Although it's async, resolving does not promise the item trees are updated.
+     */
+    async unregisterColumn(dataKey) {
+        const success = removeColumn(dataKey);
+        if (!success) {
+            return false;
+        }
+        await this._reset();
+        return true;
+    }
+
+    // TODO: add cell renderer registration
+
+    /**
+     * Get all registered columns
+     */
+    get columns() {
+        return getColumns();
+    }
+
+    /**
+     * Get column(s) that matches the properties of option
+     * @param {Partial.<ItemTreeColumnOption> | Partial.<ItemTreeColumnOption>[]} option - An option or array of options to match
+     * @returns {ItemTreeColumnOption[]}
+     */
+    getColumn(option) {
+        return getColumn(option);
+    }
+
+    /**
+     * Reset the item trees to update the columns
+     * @private
+     */
+    async _reset() {
+        await Zotero.DB.executeTransaction(async function () {
+            Zotero.Notifier.queue(
+                'refresh',
+                'itemtree',
+                [],
+                {},
+            );
+        });
+    }
+
+    /**
+     * Unregister all columns registered by a plugin
+     * @param {string} pluginID - Plugin ID
+     * @async
+     * @private
```

Comment:
No need for this — we'll use a [plugin](https://stackoverflow.com/questions/43324181/jsdoc-make-all-underscore-methods-private) to automatically make all underscore-prefixed functions private.

Reply:
Well, I guess if we're relying on JSDoc comments for TypeScript support as well then it might be useful, unless there's a way to configure that for TypeScript as well.